debian/manpages/ibterm.1
.TH ipterm 1 "February 18, 2018"

ibterm help info -

Summary:
An interactive terminal program for sending commands to a device over an 
IEEE488 general purpose instrument bus and printing the responses.

.SH OPTIONS
.TP
.B \-d <device primary address (pad)>                (required, no default)
.TP
.B \-m <minor>                                       (optional, default=0)
.TP
.B \-s <secondary address of device (sad)>           (optional, default=0)
.TP
.B \-i <assert EIO with last byte sent (eoi) flag>   (optional, default=1)
.TP
.B \-e <ASCII code of end of string (eos) character> (optional, default=0)
.TP
.B \-r <terminate read on eos character (reos) flag> (optional, default=0)
.TP
.B \-b <binary eos character match mode (bin)  flag> (optional, default=0)
.TP
.B \-x <assert EOI when tranmitting eos (xeos) flag> (optional, default=0)
.TP
.B \-t <timeout>       (optional, default=10 i.e. 300milliseconds)
.TP
.B \-p <prompt string> (optional, default="ibterm>")
.TP
.B \-f <history file>  (optional, default=".ibterm_hist_<pad>")
.TP
.B \-N No automatic read on device, enter return at prompt to read.
.TP
.B \-X forces hexadecimal output.
.TP
.B \-h prints this help info and exits.

Operation:
  loop:
    Prompt on stdout
    Read a line of text from stdin
    Write the text (if any) to the device at pad
    If -N is not set, or no text was entered
        Attempt to read response from the device
        If no response is received before timeout go to loop
        else print the response on stdout
    Go back to loop.

Notes:
  To quit the programme enter the EOF character (Ctrl-d) at the prompt.
  For interactivity, timeout should not be greater than 13 i.e. 10 secs.
  The timeout values are documented under the ibtmo() entry of the 
  section Linux-GPIB Reference: http://linux-gpib.sourceforge.net/doc_html
  A device read can always be triggered by hitting <Enter> at the prompt.
  Interrupting the programme while reading from the device may cause hangs.

Also:
  See the readline(3) manpage for editing input and searching history.

Usage:
  /usr/bin/ibterm -d pad \
         [-m minor] [-s sad] [-i eoi] [-e eos] [-r reos] [-b bin] [-x xeos] \
         [-t timeout] [-p prompt] [-f history_file] [-N] [-X]


ChangeLog
2023-07-19 13:31  dpenkler

	* [r2057] ChangeLog, README.txt: Update README.txt and ChangeLog
	  for 4.3.6

2023-07-19 13:29  dpenkler

	* [r2056] linux-gpib-user/doc/linux-gpib.sgml: Add note in docs on
	  NI GPIB-USB-HS clones


2023-07-19 12:11  dpenkler

	* [r2055] linux-gpib-kernel/drivers/gpib/include/version.h,
	  linux-gpib-user/configure.ac: Bump versions

2023-07-13 11:25  dpenkler

	* [r2054] linux-gpib-kernel/drivers/gpib/ines/ines_cs.c: Fix bug
	  #83 pcmcia drver struct name initialisation

2023-06-06 13:55  dpenkler

	* [r2053] linux-gpib-kernel/compat/include/linux/device.h,
	  linux-gpib-kernel/drivers/gpib/sys/osinit.c: Add class_create
	  compat call for linux 6.4

2023-06-06 13:27  dpenkler

	* [r2052] linux-gpib-kernel/drivers/gpib/gpio/gpib_bitbang.c:
	  Remove unreferenced #defines

2023-06-06 06:38  dpenkler

	* [r2051] linux-gpib-kernel/drivers/gpib/gpio/gpib_bitbang.c:
	  Corrections for lost interrutps from Marcello Carla'

	  - Use LEVEL interrupts for first NRFD and DAV interrupts
	  switching to EDGE thereafter

	  - Remove DAV level check in bb_read

	  - Reinstate SET_DIR_READ in set_atn

	  - Remove SET_READ_HOLDOFF

	  - Remove remaining UDELAY calls

2023-04-19 08:46  dpenkler

	* [r2050] linux-gpib-kernel/drivers/gpib/ni_usb/ni_usb_gpib.c:
	  Avoid async IO race conditions in ni_usb_gpib
	  Fix for bug report #82
	  Add addressed_transfer mutex protection around
	  bulk send and receive pairs for command_chunk,
	  take_control and goto standby.

2023-04-07 10:08  dpenkler

	* [r2049] linux-gpib-kernel/drivers/gpib/gpio/gpib_bitbang.c:
	  Workaround for lost interrutps
	  - Use irq_set_irq_type instead of enable/disable_irq.

	  - Add delay before spin_lock in bb_write when ATN is asserted.

	  - Terminate write on NRFD interrupt for command writes
	  to reduce spurious interrupt when idle conditions.

	  - Remove assertion of ATN in command write as it is done
	  by the library via take control.

	  - Terminate on NDAC for non command writes.

	  - Use SET_READ_HOLDOFF instead of SET_DIR_READ in ATN
	  transition to LACS to avoid DAV and EOI glitches with
	  ATN asserted.

	  - Remove UDELAY in SET_DIR_READ/WRITE and reorder line
	  transitions to reduce glitches.

	  - Gather irq statistics and print them on module unload.

2022-12-07 10:20  dpenkler

	* [r2048] linux-gpib-user/doc/linux-gpib.sgml: Update docs for
	  gpio_bitbang: pin_map and sn7516x_used

2022-12-07 10:19  dpenkler

	* [r2047] linux-gpib-kernel/drivers/gpib/gpio/gpib_bitbang.c: From
	  Marcello Carla' Module parameters: pin_map and sn7516x_used

2022-11-09 13:03  dpenkler

	* [r2046] linux-gpib-kernel/drivers/gpib/gpio/gpib_bitbang.c:
	  Cleanup naming and error handling

2022-11-09 11:24  dpenkler

	* [r2045] linux-gpib-user/doc/linux-gpib.sgml: Update docs for
	  gpib_bitbang board_id module parameter

2022-11-09 10:41  dpenkler

	* [r2044] linux-gpib-kernel/drivers/gpib/gpio/gpib_bitbang.c: Apply
	  patch from Marcello Carla' to unify board type mechanism

2022-11-08 08:15  dpenkler

	* [r2043] linux-gpib-kernel/drivers/gpib/gpio/gpib_bitbang.c: Add
	  patch for lsi-gpib4pi-1.1 shield from Vladimir Vassilev

2022-10-14 16:14  dpenkler

	* [r2042] linux-gpib-kernel/drivers/gpib/gpio/gpib_bitbang.c: Go
	  back to using spin_lock_irqsave/irqrestore
	  - Move enable/disable_irq out of spinlock section
	  - Associate meaningful names with irq's for /proc/interrupts

2022-09-29 11:54  dpenkler

	* [r2041] linux-gpib-kernel/drivers/gpib/ni_usb/ni_usb_gpib.c:
	  Revert commit r2038

2022-09-29 11:29  dpenkler

	* [r2040] linux-gpib-kernel/drivers/gpib/gpio/gpib_bitbang.c:
	  Updates from Marcello Carla'
	  - Fix resource management for attach fail cases
	  - Use lighter spin_(un)lock_irq() in place of
	  spin_lock_irq(restore)save() in bb_read/bb_write()
	  - Debug output clean-up

2022-09-07 15:35  dpenkler

	* [r2039] linux-gpib-kernel/drivers/gpib/ni_usb/ni_usb_gpib.c: Add
	  message for addressing error

2022-08-28 08:10  dpenkler

	* [r2038] linux-gpib-kernel/drivers/gpib/ni_usb/ni_usb_gpib.c: Add
	  ready-response option for suspected NI-USB-HS clone

2022-08-26 14:24  dpenkler

	* [r2037] linux-gpib-user/lib/async.c, linux-gpib-user/lib/ibRd.c,
	  linux-gpib-user/lib/ibWrt.c, linux-gpib-user/lib/ib_internal.h:
	  Fix for bug #81

	  -- ibrda and ibwrta routines do not use the timeout value valid
	  when
	  the call is made
	  Fix: Copy the timeout value to the async gpio aio arg and use
	  it in the spawned thread.
	  Thanks to Michael Katzmann for pointing out the bug and providing
	  a test programme.

2022-08-26 12:33  dpenkler

	* [r2036]
	  linux-gpib-kernel/drivers/gpib/agilent_82350b/agilent_82350b_read.c:
	  Remove debug printk

2022-08-26 12:28  dpenkler

	* [r2035]
	  linux-gpib-kernel/drivers/gpib/agilent_82350b/agilent_82350b.h,
	  linux-gpib-kernel/drivers/gpib/agilent_82350b/agilent_82350b_init.c,
	  linux-gpib-kernel/drivers/gpib/agilent_82350b/agilent_82350b_interrupt.c,
	  linux-gpib-kernel/drivers/gpib/agilent_82350b/agilent_82350b_read.c,
	  linux-gpib-kernel/drivers/gpib/agilent_82350b/agilent_82350b_write.c:
	  Fixed the agilent_82350b accel write block

	  -- When reading the Bus Status Register whether directly from the
	  TMS9914 (BSR) or via the 82350 register set (offset 0x13) before
	  doing the fifo write the write blocks and nothing is sent. When
	  using ibwrt there is no line_status call between sending the
	  command bytes and the data bytes -> so it always works. A
	  subsequent unaddressed write after ibwrt() fails because of the
	  line_status call from the general_library_exit ibstatus() call.
	  However if you systematically write the first byte of each buffer
	  using PIO mode and the rest using the fifos it works just fine.
	  BO and BI interrupts are now disabled in write and read
	  respectively.
	  Set the fifo size for both read and write to 0x8000

2022-08-26 11:18  dpenkler


linux-gpib-user/examples/ibterm.c
/***************************************************************************
                                 ibterm.c
                            ------------------

 Another example program which uses the linux gpib C library:

   An interactive terminal program for sending commands to an instrument
   and printing its response.

    copyright : (C) 2013 by Dave Penkler
    email     : dpenkler@gmail.com
 ***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/

#include <ctype.h>
#include <stdio.h>
#include <string.h>
#include <malloc.h>
#include <stdlib.h>
#include <getopt.h>
#ifdef READLINE
#include <readline/readline.h>
#include <readline/history.h>
#endif
#include "gpib/ib.h"

#define PROMPT "ibterm>"
#define DEVICE_BUFFER_SIZE 2048
#define false 0
#define true  1

/* Global variable declarations */
static char * mPrompt   = PROMPT;        // Prompt string
static char * mProg;                     // Programme name
static char * mHist     = (char *)NULL;  // History filename pointer
static int mAutoRead    = true;          // Automatically read from device
static int mHex         = false;         // Force hex output flag
static int mUnTUnL      = false;         // send untalk/unlisten
static int devdesc  = -1;  // device descriptor
static int minor    = 0;   // gpib driver major
static int pad      = -1;  // device primary bus address
static int sad      = 0;   // device secondary bus address
static int send_eoi = 1;   // Assert eoi with last byte in ibwrt
static int eos_mode = 0;   // No end of string processing
static int timeout  = 10;  // 300 milliseconds
static char errmes[256];   // error message buffer

static const char * help_string =
  "ibterm help info -\n"
  "\nSummary:\n"
  "An interactive terminal program for sending commands to a device over an \n"
  "IEEE488 general purpose instrument bus and printing the responses.\n"
  "\nOptions:\n"
  " -d <device primary address (pad)>                (required, no default)\n"
  " -m <minor>                                       (optional, default=0)\n"
  " -s <secondary address of device (sad)>           (optional, default=0)\n"
  " -i <assert EIO with last byte sent (eoi) flag>   (optional, default=1)\n"
  " -e <ASCII code of end of string (eos) character> (optional, default=0)\n"
  " -r <terminate read on eos character (reos) flag> (optional, default=0)\n"
  " -b <binary eos character match mode (bin)  flag> (optional, default=0)\n"
  " -x <assert EOI when tranmitting eos (xeos) flag> (optional, default=0)\n"
  " -t <timeout>       (optional, default=10 i.e. 300milliseconds)\n"
  " -p <prompt string> (optional, default=\"" PROMPT "\")\n"
#ifdef READLINE
  " -f <history file>  (optional, default=\".ibterm_hist_<pad>\")\n"
#endif
  " -u Send Untalk/Unlisten after each read and write\n"
  " -N No automatic read on device, enter return at prompt to read.\n"
  " -X forces hexadecimal output.\n"
  " -h prints this help info and exits.\n"
  "\nOperation:\n"
  "  loop:\n"
  "    Prompt on stdout\n"
  "    Read a line of text from stdin\n"
  "    Write the text (if any) to the device at pad\n"
  "    If -N is not set, or no text was entered\n"
  "        Attempt to read response from the device\n"
  "        If no response is received before timeout go to loop\n"
  "        else print the response on stdout\n"
  "    Go back to loop.\n"
  "\nNotes:\n"
  "  To quit the programme enter the EOF character (Ctrl-d) at the prompt.\n"
  "  For interactivity, timeout should not be greater than 13 i.e. 10 secs.\n"
  "  The timeout values are documented under the ibtmo() entry of the \n"
  "  section Linux-GPIB Reference: http://linux-gpib.sourceforge.net/doc_html\n"
  "  A device read can always be triggered by hitting <Enter> at the prompt.\n"
  "  Interrupting the programme while reading from the device may cause hangs.\n"
#ifdef READLINE
  "\nAlso:\n"
  "  See the readline(3) manpage for editing input and searching history.\n"
#endif
  ;

static const char * usage_options =
  " -d pad \\\n"
  "         [-m minor] [-s sad] [-i eoi] [-e eos]"
  " [-r reos] [-b bin] [-x xeos] \\\n"
  "         [-t timeout] [-p prompt]"
#ifdef READLINE
  " [-f history_file]"
#endif
  " [-u] [-N] [-X] [-h]\n";

#define EMES(var) fputs(var,stderr)

static void abend(const char * mess) {
  EMES(mess);
  EMES(mProg);
  EMES(": Aborted\n");
  if (devdesc >= 0) ibonl(devdesc,0);
  exit(1);
}

static void usage(int abort) {
  EMES("\nUsage:\n  ");
  EMES(mProg);
  EMES(usage_options);
  if (abort) {
    sprintf(errmes,"Try '%s -h' for more information.\n",mProg);
    abend(errmes);
  }
  exit(1);
}

/* Print buffer contents in parallel hexadecimal and ASCII */
static const char * hexdig = "0123456789ABCDEF";
static char * template =
  "12345678 |                         | "
             "                        |                 ";
void prhex(unsigned char * buf, int len) {
  int i,j,k,l,m;
  char *c,ob[80];

  for (i=0;i<len;i++)  {
    if ((i % 16) == 0) {
      k = 8; l = 63;
      strcpy(ob,template);
      for (j=0,c=&ob[8],m=i;j<8;m >>= 4,j++)
	*--c = hexdig[m & 0xf];
    }
    k += 1 + 2*((i % 8) == 0);
    ob[k++] = hexdig[(buf[i] & 0xf0)>>4];
    ob[k++] = hexdig[(buf[i] & 0xf)];
    ob[l++] = (buf[i] < 32 || buf[i] > 127) ? ' ' : buf[i];
    if (k == 60) { puts(ob); k = 0; }
   }
  if (k) puts(ob);
}

#define CHECK_FLAG(var, flag)			\
if (var != 0 && var != 1)  abend(#flag " flag must be 1 or 0.\n");

#define CHECK_ADDR(var)		  \
if (var < 0 || var > 30)  abend(#var " must be between 0 and 30.\n");

#define CHECK_SADDR(var) 	\
if (var < 0x60 || var > 0x7f)  abend("linux-gpib requires the secondary address to be offset by 96,\n   that is sad must be between 96 and 127.\n");

void parse_options(int argc, char ** argv) {
  int eos_char  = 0;   // End of string character
  int reos_mode = 0;   // Don't terminate read on eos_char
  int bin_mode  = 0;   // Match only 7 bits of eos_char
  int xeos_mode = 0;   // Don't assert EOI when sending eos_char
  int c;

  mProg = argv[0];

  while ((c = getopt (argc, argv, "d:m:s:i:e:r:b:x:t:f:p:uNXh")) != -1)
    switch (c)  {
    case 'd': pad       = atoi(optarg);    break;
    case 'm': minor     = atoi(optarg);    break;
    case 's': sad       = atoi(optarg);    break;
    case 'i': send_eoi  = atoi(optarg);    break;
    case 'e': eos_char  = atoi(optarg);    break;
    case 'r': reos_mode = atoi(optarg);    break;
    case 'b': bin_mode  = atoi(optarg);    break;
    case 'x': xeos_mode = atoi(optarg);    break;
    case 't': timeout   = atoi(optarg);    break;
    case 'f': mHist     = optarg;          break;
    case 'p': mPrompt   = optarg;          break;
    case 'u': mUnTUnL   = true;            break;
    case 'N': mAutoRead = false;           break;
    case 'X': mHex      = true;            break;
    case 'h':
      EMES(help_string);
      usage(0);
      break;
    default: usage(1);
    }
  if (pad == -1) {
    EMES("No primary device address (pad) specified!\n");
    usage(1);
  }
  CHECK_ADDR(pad);
  if (sad) CHECK_SADDR(sad);
  CHECK_FLAG(send_eoi, eoi);
  CHECK_FLAG(reos_mode,reos);
  CHECK_FLAG(bin_mode, bin);
  CHECK_FLAG(xeos_mode,xeos);
  if (eos_char < 0 || eos_char > 255) {
    abend("eos character must be between 0 and 255,\n");
  }
  if (!bin_mode && eos_char > 127) {
    EMES("Warning eos is 8 bits but compares are set to 7.\n");
  }
  if (timeout < 1 || timeout > 15) abend("<timeout> must be between 1 and 15.");
  eos_mode = eos_char;
  if (reos_mode) eos_mode |= REOS;
  if (bin_mode)  eos_mode |= BIN;
  if (xeos_mode) eos_mode |= XEOS;
  if (mHist == NULL) {
    mHist = malloc(32);
    sprintf(mHist,".ibterm_hist_%02d",pad);
  }
}

static void showError(char * mess) {
  fprintf(stderr,"ibterm error: %s\n", mess);
  fprintf(stderr," - %s\n", gpib_error_string(ThreadIberr()));
  if (iberr==EDVR)
	  fprintf(stderr,"\t%s\n",strerror(ibcnt));
}

#ifndef READLINE
#define READ_BUF_SIZE 2048
char * readline(char * prompt) {
  char * buf, * ret;
  int len;
  fwrite(prompt,1,strlen(prompt),stdout);
  fflush(stdout);
  if (!(buf = malloc(READ_BUF_SIZE))) {
    abend("no memory");
  }
  ret = fgets(buf,READ_BUF_SIZE,stdin);
  if (ret) {
    len = strlen(ret);
    if (ret[len-1] == '\n') ret[len-1] = 0;
  } else  free(buf);
  return ret;
}
  void read_history(char * buf) {};
  void add_history(char * buf) {};
  void write_history(char * buf) {};
#endif

int main (int argc, char ** argv) {
  char * line;
  int i, devdatalen,datasent;
  unsigned char devbuf[DEVICE_BUFFER_SIZE];
  int printable;

  parse_options(argc, argv);

  printf("Attempting to open /dev/gpib%i\n"
	 "pad = %d, sad = %d, timeout = %d, send_eoi = %d, eos_mode = 0x%04x\n",
	 minor,pad,sad,timeout,send_eoi,eos_mode);
  putenv("IB_NO_ERROR=1"); // we check for our own errors
  devdesc = ibdev(minor, pad, sad, timeout, send_eoi, eos_mode);
  if (devdesc < 0) {
    showError("open failed");
    abend("ibdev error\n");
  }

  if (mUnTUnL) {
    if(ERR & ibconfig(devdesc, IbcUnAddr, 1))
      EMES("Could not set IbcUnAddr\n");
    ibask(devdesc, IbaUnAddr, &i);
    printf("IbaUnAddr %d\n",i);
  }

  read_history(mHist);

  /* prompt user - read term - write device - read device - print term loop */
  while (line = readline(mPrompt)) { /* prompt  and read from user */

    /* write to device */
    if (*line) { /* send to device only if we got something */
      if (ibwrt(devdesc,line,strlen(line)) & ERR ) {
	sprintf(errmes,"Unable to write to device at pad %d\n",pad);
	showError(errmes);
	free(line);
	continue;
      }
      add_history (line);
      datasent = true;
    } else  datasent = false;

    free(line); /* return buffer allocated by readline */

    if (mAutoRead || !datasent) {   /* read response if any from device */
      if (ibrd(devdesc,devbuf,DEVICE_BUFFER_SIZE) & ERR) {
	if (!(ThreadIbsta() & TIMO)) {
	  sprintf(errmes,"Failed during read from device at pad %d\n",pad);
	  showError(errmes);
	}
	continue;
      }
      devdatalen = ThreadIbcntl();
    } else continue;

    /* print response */
    if (!mHex) {
      printable  = true;
      for(i = 0; i < devdatalen; ++i) {
	if (!isprint(devbuf[i]) && !isspace(devbuf[i])){
	  printable = false;
	  break;
	}
      }
    } else printable = false;
    if (printable) {
      fwrite(devbuf,1,devdatalen,stdout); /* print response as ASCII string */
      /* but don't print an extra newline if it ends with LF or LF CR */
      if ((devdatalen > 2 &&
	   devbuf[devdatalen-2] != '\n' ||  devbuf[devdatalen-1] != '\r') &&
	  devbuf[devdatalen-1] != '\n') putchar('\n');
    } else {
      prhex(devbuf,devdatalen); /* print contents with hex and ascii */
    }
  }

  /* Finish up */
  ibonl(devdesc, 0);   /* free up gpib library resources */
  write_history(mHist);
  puts("\nibterm: Done.");
  exit(0);
}
